### PR TITLE
Limit data get

### LINKED ISSUE #267: Implement data download and plotting limits
Maximum data download should be 1 year of data, Will need to profile loading the max timespan and limit the update graph and download function separately if it becomes an issue.

### PR DESCRIPTION
**Pull request by lboeman**
Adds javascript for limitting the amount of data that cal be plotted/downloaded through the dashboard. Right now allows for downloading 1 year of data, and I plan to implement limiting plotting data takes on average < 10s to load across interval lengths. This feature requires loading the metadata of the current object into a js variable which will probably be useful elsewhere.
closes #267

### CODE REVIEW

**Comment by lboeman** [comment]
Now adds javascript and python handling for exceeding the maximum plotting or data request limits. Handles a few cases differently:

- Short interval lengths where the plotting datapoint limit is reached before one year of data. I've set this limit to be 93 days of 1 minute data, or 133920 data points. This amount of data loads consistently at less than 10 seconds across different interval lengths. When the user enters a start/end that exceeds the data point limit, the "Update graph" button is disabled and a message is displayed.
Screenshot from 2020-06-01 14-59-50

- When over a year of data is entered both "Update graph" and "Download data" are disabled and a warning is displayed.
Screenshot from 2020-06-01 15-00-05

- The insert_plot function now checks that limits are not exceeded, and displays a warning in place of a plot. This handles a case where the query parameters are set from a source other than the update/download form. Either a user adding them manually to the url or Flask's url_for function. It provides a limit in terms of days and datapoints for the user.
Screenshot from 2020-06-01 15-49-26

File: sfa_dash/blueprints/base.py [hunk 1]
```diff
@@ -19,34 +20,64 @@ def insert_plot(self, uuid, start, end):
...
-        except DataRequestException as e:
-            if e.status_code == 422:
-                self.temp_args.update({'warnings': e.errors})
+        # determine limit by number of datapoints in the maximum allowable
+        # time range based on interval length.
+        timerange_limit = current_app.config['MAX_DATA_RANGE_DAYS']
+        interval_length = self.metadata['interval_length']
+        max_pts = timerange_limit / pd.Timedelta(f'{interval_length} minutes')
+        if max_pts > current_app.config['MAX_PLOT_DATAPOINTS']:
+            max_pts = current_app.config['MAX_PLOT_DATAPOINTS']
```
Comment (alorenzo175) [inline sfa_dash/blueprints/base.py:37]:
```suggestion
        max_pts = min(
            timerange_limit / pd.Timedelta(f'{interval_length} minutes'),
            current_app.config['MAX_PLOT_DATAPOINTS']
        )
```

File: sfa_dash/blueprints/base.py [hunk 2]
```diff
@@ -19,34 +20,64 @@ def insert_plot(self, uuid, start, end):
...
-                    'bokeh_script': script_plot[0]
-                })
+            allowable_days = pd.Timedelta(f"{max_pts*interval_length} minutes")
+            self.temp_args.update({
+                'plot': '<div class="alert alert-warning">Too many datapoints '
+                        'to display. The maximum number of datapoints to plot '
+                        f'is {max_pts}. This amounts to {allowable_days.days} '
+                        f'days and {allowable_days.components.hours} hours of '
+                        f'data. The the requested data contains {total_points}'
+                        '.</div>'
```
Comment (alorenzo175) [inline sfa_dash/blueprints/base.py:79]:
```suggestion
                        ' datapoints.</div>'
```

File: sfa_dash/static/js/timerange-handling.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
+MAX_DOWNLOAD_DAYS = 366;
+MAX_PLOT_POINTS = 1440 * 93; // approximately 3 months of 1 minute data
```
Comment (alorenzo175) [inline sfa_dash/static/js/timerange-handling.js:2]: no way these can be set from the app config?

File: sfa_dash/blueprints/base.py [hunk 4]
```diff
@@ -198,6 +234,19 @@ def flash_api_errors(self, errors):
...
+    def safe_metadata(self):
+        """Creates a copy of the metadata attribute without the
+        `extra_parameters` key.
+        """
+        safe_metadata = deepcopy(self.metadata)
+        safe_metadata.pop('extra_parameters', None)
+        if 'site' in safe_metadata:
+            safe_metadata['site'].pop('extra_parameters', None)
+        if 'aggregate' in safe_metadata:
+            safe_metadata['aggregate'].pop('extra_parameters', None)
```
Comment (alorenzo175) [inline sfa_dash/blueprints/base.py:246]: hmm would it be safer to recurse through and pop any 'extra_parameters' keys found?

File: sfa_dash/templates/data/asset.html [hunk 5]
```diff
@@ -1,7 +1,12 @@
 {% extends "dash/data.html" %}
 {% import "forms/form_macros.jinja" as form %}
+{% block extra_scripts %}
+{% if metadata is defined %}
+  <script>var metadata = JSON.parse('{{ metadata | tojson }}');console.log('extra loaded');</script>
```
Comment (alorenzo175) [inline sfa_dash/templates/data/asset.html:5]: what's the log for?

File: sfa_dash/static/js/timerange-handling.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    if (days > 0){
+        // limit maximum amount of data to download
+        if (days > MAX_DOWNLOAD_DAYS){
+            // disable download and plot update
+            $('#download-submit').attr('disabled', true);
+            $('#plot-range-adjust-submit').attr('disabled', true);
+
+            insertWarning(
+                'Maximum timerange exceeded',
+                'Maximum of one year of data may be requested.'
```
Comment (alorenzo175) [inline sfa_dash/static/js/timerange-handling.js:52]:
```suggestion
                'Maximum of one year of data may be requested at once.'
```

**Pull request merged**